### PR TITLE
APPS-3522: Google billing library

### PR DESCRIPTION
**Pull request by rostikjoystick**
**YouTrack task**: [#APPS-3522](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-3522),  [#APPS-3490](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-3490)

**Description List**:
* Remove Checkout library, add Google Billing
* Handle buy action

### CODE REVIEW

File: app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt [hunk 1]
```diff
@@ -1,20 +1,62 @@
...
+    private val coursePaymentsRepository: CoursePaymentsRepository,
+
+    private val sharedPreferenceHelper: SharedPreferenceHelper,
+
+    private val courseRepository: CourseRepository,
+    private val lessonRepository: LessonRepository,
+
+    @EnrollmentCourseUpdates
+    private val enrollmentSubject: PublishSubject<Course>,
+    private val gson: Gson
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt:53]: Gson not used

File: app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt [hunk 2]
```diff
@@ -23,8 +65,81 @@ constructor(
...
+            }
+
+    private fun updateCourseAfterEnrollment(courseId: Long): Completable =
+        userCoursesInteractor.addUserCourse(courseId)
+            .andThen(lessonRepository.removeCachedLessons(courseId))
+            .andThen(courseRepository.getCourse(courseId, sourceType = DataSourceType.REMOTE, allowFallback = false).toSingle())
+            .doOnSuccess(enrollmentSubject::onNext) // notify everyone about changes
+            .ignoreElement()
+
+    private fun getCurrentProfileId(): Single<Long> =
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt:138]: Let use `profileRepository`

File: app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt [hunk 3]
```diff
@@ -23,8 +65,81 @@ constructor(
...
+                if (payment.status == CoursePayment.Status.SUCCESS) {
+                    Completable.complete()
+                } else {
+                    Completable.error(CoursePurchaseVerificationException())
+                }
+            }
+            .andThen(updateCourseAfterEnrollment(courseId))
+            .andThen(billingRepository.consumePurchase(purchase))
+
+    fun restorePurchase(courseId: Long, skuId: String): Completable =
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt:97]: Will it work if user bought course with promo code in past? Or course price was changed.

File: app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt [hunk 4]
```diff
@@ -23,8 +65,81 @@ constructor(
...
+    private fun restorePurchase(courseId: Long, sku: SkuDetails): Completable =
+        zip(
+            getCurrentProfileId()
+                .toMaybe(),
+            billingRepository
+                .getAllPurchases(BillingClient.SkuType.INAPP, listOf(sku.sku))
+                .maybeFirst()
+        )
+            .filter { (profileId, purchase) ->
+                val obfuscatedAccountId = profileId.toString().hashCode().toString()
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt:111]: code duplication

File: app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt [hunk 5]
```diff
@@ -23,8 +65,81 @@ constructor(
                     Single.just(PromoCodeSku.EMPTY)
                 } else {
                     lightSkuRepository
-                        .getLightInventory(ProductTypes.IN_APP, listOf(mobileTier.promoTier), dataSourceType = DataSourceType.REMOTE)
+                        .getLightInventory(BillingClient.SkuType.INAPP, listOf(mobileTier.promoTier), dataSourceType = DataSourceType.REMOTE)
                         .map { lightSku -> PromoCodeSku(promoCodeName, lightSku.firstOrNull()) }
                 }
             }
+
+    fun launchPurchaseFlow(courseId: Long, skuId: String): Single<Pair<CoursePurchaseObfuscatedParams, SkuDetails>> =
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt:73]: `launchPurchaseFlow` actually not launching anything

File: app/src/main/java/org/stepik/android/presentation/course/CoursePresenter.kt [hunk 6]
```diff
@@ -209,21 +197,25 @@ constructor(
...
                 enrollCourse()
 
             EnrollmentState.NotEnrolledWeb ->
                 openCoursePurchaseInWeb()
 
-            is EnrollmentState.NotEnrolledInApp ->
-                purchaseCourse()
+            is EnrollmentState.NotEnrolledMobileTier -> {
+                // no op
+//                purchaseCourse(enrollmentState.standardLightSku, headerData.deeplinkPromoCodeSku)
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/presentation/course/CoursePresenter.kt:217]: Why not to start purchase flow?

File: app/src/main/java/org/stepik/android/presentation/course/CoursePresenter.kt [hunk 7]
```diff
@@ -276,7 +268,7 @@ constructor(
             .subscribeOn(backgroundScheduler)
             .observeOn(mainScheduler)
             .subscribeBy(
-                onNext  = { state = CourseView.State.CourseLoaded(it); continueLearning(); resolveCourseShareTooltip(it) },
+                onNext  = { state = CourseView.State.CourseLoaded(it); resolveCourseShareTooltip(it) },
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/presentation/course/CoursePresenter.kt:271]: Seems that it broke free and web enrolments

File: app/src/main/java/org/stepik/android/presentation/course/CoursePresenter.kt [hunk 8]
```diff
@@ -318,91 +310,6 @@ constructor(
...
-
-        val sku = (headerData.stats.enrollmentState as? EnrollmentState.NotEnrolledInApp)
-            ?.skuWrapper
-            ?.sku
-            ?: return
-
-        val checkout = this.uiCheckout
-            ?: return
-
-        schedulePurchaseReminder()
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/presentation/course/CoursePresenter.kt:379]: seems that we lost it

File: app/src/main/java/org/stepik/android/presentation/course_purchase/dispatcher/CoursePurchaseActionDispatcher.kt [hunk 9]
```diff
@@ -26,6 +37,21 @@ constructor(
...
+            .subscribeOn(backgroundScheduler)
+            .observeOn(mainScheduler)
+            .subscribeBy(
+                onNext = { (billingResult, purchases) ->
+                    if (billingResult.responseCode == BillingClient.BillingResponseCode.OK && purchases?.firstOrNull() != null) {
+                        onNewMessage(CoursePurchaseFeature.Message.PurchaseFlowBillingSuccess(purchases.first()))
+                    } else {
+                        onNewMessage(CoursePurchaseFeature.Message.PurchaseFlowBillingFailure(BillingException(billingResult.responseCode).toEnrollmentError()))
+                    }
+                    Timber.d("Action dispatcher; Billing Result - Debug message: ${billingResult.debugMessage} Response: ${billingResult.responseCode}; Purchases - $purchases")
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/presentation/course_purchase/dispatcher/CoursePurchaseActionDispatcher.kt:51]: logs

File: app/src/main/java/org/stepik/android/view/course/ui/delegates/CourseHeaderDelegate.kt [hunk 10]
```diff
@@ -348,27 +334,9 @@ constructor(
...
-                        courseHeaderData.course,
-                        courseHeaderData.stats,
-                        notEnrolledMobileTierState.standardLightSku,
-                        promoCodeSku,
-                        courseHeaderData.course.isInWishlist
-                )
-                coursePurchaseFlowAction(coursePurchaseData)
+        val coursePurchaseData = resolveCoursePurchaseData(courseHeaderData)
+        if (coursePurchaseData != null) {
+            coursePurchaseFlowAction(coursePurchaseData, false)
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/view/course/ui/delegates/CourseHeaderDelegate.kt:339]: Lost buy course pressed analytics

File: billing/src/main/java/org/stepik/android/domain/billing/repository/BillingRepository.kt [hunk 11]
```diff
@@ -1,41 +1,26 @@
...
+import com.android.billingclient.api.SkuDetails
 import io.reactivex.Completable
 import io.reactivex.Maybe
 import io.reactivex.Single
-import org.solovyev.android.checkout.Purchase
-import org.solovyev.android.checkout.Sku
 import ru.nobird.android.domain.rx.maybeFirst
 
 interface BillingRepository {
-    /**
```
Comment (eadm) [inline billing/src/main/java/org/stepik/android/domain/billing/repository/BillingRepository.kt:11]: docs 😢

File: billing/src/main/java/org/stepik/android/view/injection/billing/BillingDataModule.kt [hunk 12]
```diff
@@ -10,8 +10,10 @@ import org.stepik.android.remote.billing.BillingRemoteDataSourceImpl
 @Module
 abstract class BillingDataModule {
     @Binds
-    internal abstract fun bindBillingRepository(billingRepositoryImpl: BillingRepositoryImpl): BillingRepository
+    @BillingSingleton
+    internal abstract fun bindBillingNewRepository(billingNewRepositoryImpl: BillingRepositoryImpl): BillingRepository
```
Comment (eadm) [inline billing/src/main/java/org/stepik/android/view/injection/billing/BillingDataModule.kt:14]: `new`

File: billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt [hunk 13]
```diff
@@ -1,37 +1,54 @@
...
-            .also(Checkout::start)
+    @BillingSingleton
+    internal fun provideBillingClient(
+        context: Context,
+        purchaseListenerBehaviorRelay: BehaviorRelay<Pair<BillingResult, List<Purchase>?>>
+    ): BillingClient =
+        BillingClient
+            .newBuilder(context)
+            .setListener { billingResult, mutableList ->
+                Log.d("APPS", "Result: ${billingResult.debugMessage} Code: ${billingResult.responseCode}")
```
Comment (eadm) [inline billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt:31]: logs

File: app/src/main/java/org/stepik/android/presentation/course_purchase/dispatcher/CoursePurchaseActionDispatcher.kt [hunk 14]
```diff
@@ -8,16 +12,23 @@ import org.stepic.droid.di.qualifiers.BackgroundScheduler
...
 import org.stepik.android.presentation.course_purchase.CoursePurchaseFeature
+import org.stepik.android.view.injection.billing.BillingSingleton
 import ru.nobird.android.presentation.redux.dispatcher.RxActionDispatcher
+import timber.log.Timber
 import javax.inject.Inject
 
 class CoursePurchaseActionDispatcher
 @Inject
 constructor(
+    @BillingSingleton
```
Comment (eadm) [inline app/src/main/java/org/stepik/android/presentation/course_purchase/dispatcher/CoursePurchaseActionDispatcher.kt:29]: Do we need it?

File: billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt [hunk 15]
```diff
@@ -1,37 +1,54 @@
...
+                purchaseListenerBehaviorRelay.accept(billingResult to mutableList)
+            }
+            .enablePendingPurchases()
```
Comment (eadm) [inline billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt:34]: will we receive delayed payments here?

File: billing/src/main/java/org/stepik/android/remote/billing/BillingRemoteDataSourceImpl.kt [hunk 16]
```diff
@@ -1,85 +1,105 @@
...
+    @BillingSingleton
+    private val billingClient: BillingClient,
     @RxScheduler.Main
     private val mainScheduler: Scheduler,
 
     @RxScheduler.Background
     private val backgroundScheduler: Scheduler
 ) : BillingRemoteDataSource {
-    override fun getInventory(productType: String, skuIds: List<String>): Single<List<Sku>> =
+    private fun connect(): Completable =
```
Comment (eadm) [inline billing/src/main/java/org/stepik/android/remote/billing/BillingRemoteDataSourceImpl.kt:31]: not used

File: billing/src/main/java/org/stepik/android/remote/billing/BillingRemoteDataSourceImpl.kt [hunk 17]
```diff
@@ -1,85 +1,105 @@
...
 constructor(
-    private val billing: Billing,
-
-    @SystemCheckout
-    private val checkout: Checkout,
-
+    @BillingSingleton
+    private val billingClient: BillingClient,
     @RxScheduler.Main
     private val mainScheduler: Scheduler,
```
Comment (eadm) [inline billing/src/main/java/org/stepik/android/remote/billing/BillingRemoteDataSourceImpl.kt:26]: not used

File: billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt [hunk 18]
```diff
@@ -1,37 +1,54 @@
...
+                override fun onBillingSetupFinished(billingResult: BillingResult) {
+                    Log.d("APPS", "onBillingSetupFinished - Debug message: ${billingResult.debugMessage}; Response code: ${billingResult.responseCode}")
+                }
+            })}
+
+
+    @Provides
+    @BillingSingleton
+    internal fun providePurchaseListenerRelay(): BehaviorRelay<Pair<BillingResult, List<Purchase>?>> =
+        BehaviorRelay.create()
```
Comment (eadm) [inline billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt:50]: Why behaviour relay? Do we need to store last value?

**Review by eadm — COMMENTED** [review]
- start learning button should open course and perform continue learning operation

File: app/src/main/java/org/stepic/droid/base/App.kt [hunk 19]
```diff
@@ -72,6 +74,9 @@ class App : MultiDexApplication() {
     @Inject
     internal lateinit var codeParserContainer: ParserContainer
 
+    @Inject
+    internal lateinit var billingClient: BillingClient
```
[the remark this answers is not in the document]
Comment (rostikjoystick) [reply app/src/main/java/org/stepic/droid/base/App.kt:78]: Added it here to initialize it as soon as possible, so it connects to the Google Play Store

File: billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt [hunk 20]
```diff
@@ -41,5 +58,5 @@ class BillingModule {
     @Provides
     @RxScheduler.Background
     internal fun provideBackgroundScheduler(): Scheduler =
-        Schedulers.io()
+        Schedulers.single()
```
[the remark this answers is not in the document]
Comment (rostikjoystick) [reply billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt:61]: Accident, leftover from attempts to use multiple `BillingClient` instances

File: billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt (hunk 15 above)
Comment (rostikjoystick) [reply billing/src/main/java/org/stepik/android/view/injection/billing/BillingModule.kt:34]: Yes, this listener will receive purchases with a _Pending_ status

**Review by eadm — APPROVED** [review]

**Pull request merged**